java/core/src/java/org/apache/orc/impl/RecordReaderUtils.java

```diff
@@ -182,10 +182,13 @@ public static long estimateRgEndOffset(boolean isCompressed,
                                          long streamLength) {
     // figure out the worst case last location
     // if adjacent groups have the same compressed block offset then stretch the slop
-    // by factor of 2 to safely accommodate the next compression block.
-    // One for the current compression block and another for the next compression block.
+    // by a factor to safely accommodate the next compression block.
+    // 512 is the MAX_SCOPE defined in RunLengthIntegerWriterV2.
+    // 8 is the maximum size of bytes for each value (see RunLengthIntegerWriterV2.zzBits100p).
+    // We need to calculate the maximum number of blocks by bufferSize accordingly.
+    int stretchFactor = bufferSize > 0 ? 2 + (512 * 8 - 1) / bufferSize : 2;
```

Comment:
It seems that we truncate values here. I'm wondering if we need a round-up operation for `(512 * 8 - 1) / bufferSize`? Or, is it safe to truncate?

Reply:
Not sure bufferSize can be zero/negative -- can you explain a bit more the logic here?

Reply:
Also an example with compression blocks would like help

Reply:
We are actually doing a round-up operation for `(512 * 8) / bufferSize`. We can interpret the formula as `1 (for uncompressed bytes in buffer) + Math.ceil(512 * 8 / (float) bufferSize)`. The second part can be expanded to `1 + (512 * 8 - 1) / bufferSize`. I just thought that we can avoid floating point calculation here.

Reply:
@pgaref `bufferSize == 0` if `isCompressed == false`. In other words, `bufferSize` doesn't matter in uncompressed stream.